### ISSUE TITLE
[Bug]: Qwen3-Omni seed-TTS — model speaks the task description / answers the prompt instead of reading it; ~2x worse on the vLLM 0.27 rebase

### ISSUE THREAD

**Issue by tzhouam** [vllm-project/vllm-omni#6067]
## Summary

The nightly `Omni · Accuracy Test` (`test_qwen3_omni_seed_tts_wer_bench`) fails on `dev/vllm-align` (the vLLM 0.27 rebase, PR #5976) with mean WER ~0.38–0.51 against the 0.35 gate, while `main` scores 0.27–0.29.

I reproduced the gap locally as a controlled A/B and inspected every synthesized utterance. **The failures are not audio corruption — Qwen3-Omni stops behaving as a TTS engine for a subset of prompts and instead speaks the task description, or answers the reference sentence, or babbles.** Because the WER used by the gate is uncapped, a few such clips dominate the mean.

Importantly, **this failure mode already exists on `main`; the rebase roughly doubles its frequency.** So there are two problems: a latent instruction-following bug (product) and a real regression on the rebase branch (blocking PR #5976).

## Reproduction (local, controlled A/B)

Same box (2 GPUs), same dataset (`zhaochenyang20/seed-tts-eval`, en, 1088 utterances), same harness invocation, same deploy config `vllm_omni/deploy/qwen3_omni_moe.yaml`, `--seed-tts-wer-save-items` + `SEED_TTS_WER_SAVE_AUDIO_DIR` so every utterance is scored and kept:

| stack | vLLM / torch | mean WER | median | clips with WER>1.0 |
|---|---|---|---|---|
| `origin/main` | 0.26.0 / 2.11 | **0.3047** | 0.0000 | 89 |
| `dev/vllm-align` @ `25b80bb9` | 0.27.0 / 2.13 | **0.3821** | 0.0000 | 151 |

Per-utterance comparison over the same 1088 prompts:

- **80** of main's 89 broken clips are broken on dev too (a stable hard-clip set)
- **71** clips regress (fine on main → broken on dev); only **9** improve — an 8:1 asymmetry, not a symmetric numerics reshuffle
- net **+0.0774** mean WER per clip

CI shows the same direction and a slightly larger gap (main 0.2665/0.2773/0.2788/0.2932 vs dev 0.3756/0.3826/0.3916/0.5114), so this is not specific to my hardware (I ran on L20X; CI uses H100).

## What the model actually does

Examples where `main` speaks the reference verbatim and `dev` does not (same prompt, same seed-tts request):

| reference (must be spoken verbatim) | main | dev |
|---|---|---|
| `Appearances are deceptive.` | *"Appearances are deceptive."* | *"Nezha de Siajin's speech in the same language and style as the reference audio"* |
| `Don't take any chances.` | *"Don't take any chances."* | *"Get decisive speech in the same style as the reference audio. Calm, measured, and slightly cautionary tone. Don't take any chances."* |
| `He always stutters when he's nervous, that's why he delegated the speech.` | verbatim | *"Not cause things flit in prose again, at the false as when the distairs, Miss Jew-tation the back-shadow…"* (babble) |

Longer failures are the model answering the sentence instead of reading it:

- ref `I feel sick, what should I do?` → spoken: *"I'm sorry to hear that you're feeling unwell. Here are some general steps you can take. 1. Rest… 2. Stay hydrated…"* (97 s of audio)
- ref `Please search The Best of Steely Dan: Then and Now saga.` → spoken: *"Sure. Here is the text you requested spoken in the style of the reference audio. The Best of Steely Dan… is a compilation album released by Steely Dan in 1993…"*

Failure-mode breakdown (clips with WER>0.5, share of total error):

| mode | main | dev |
|---|---|---|
| over-generation (spoke ≫ reference: answers/preambles) | 59 clips, 72% of error | 97 clips, 64% of error |
| garbled (similar length, wrong words) | 46 clips, 17% of error | 97 clips, 28% of error |
| clips that literally speak the phrase *"reference audio"* (task text leaking into speech) | **64** | **146** |

References that are questions are hit hardest on dev: mean WER **2.53** (60% above 1.0) vs 0.33 for statements.

## Why the metric amplifies this

- **WER is uncapped.** `_jiwer_wer` (`vllm_omni/benchmarks/data_modules/seed_tts_eval.py`) returns `(S+D+I)/N` over reference words. A 97 s spoken answer against a 7-word reference scores ~7.4. One such clip moves a 1088-clip mean by ~0.007; the top 10 clips alone contribute 0.055 of dev's 0.382, and clips with WER>1.0 carry **82%** of the total.
- **The talker is unseeded and stochastic** (`qwen3_omni_moe.yaml`: stage 1 `temperature: 0.9, top_k: 50`; the bench passes no seed), so the tail — and therefore the gated mean — is noisy run to run.

A tail-insensitive statistic (per-clip WER capped at 1.0, or a "fraction of clips with WER>0.5" gate) would make this job diagnose content failures instead of tracking a handful of runaway clips.

## Ruled out

- **Harness / measurement.** `git diff origin/main..HEAD` touches exactly one bench file (`vllm_omni/benchmarks/patch/patch.py`), and only with an inactive `probe_request_rate` port (default 0.0, never passed by CI) and a defensive `getattr`. The seed-tts dataset module, prompt, eval code, and `tests/e2e/accuracy/qwen3_omni/` are unchanged; `total_input_tokens` is bit-identical (158002) across branches, proving the same prompts.
- **Sampler precision.** vLLM 0.27 drops `dtype=torch.float32` from the top-p softmax in `topk_topp_sampler.py`, but `Sampler.forward` already upcasts logits to fp32 beforehand in both versions, so it is a no-op.
- **Whisper grader drift.** Same decode path in both runs; a direct A/B moved WER by −0.007.
- **Total audio volume** is *not* a reliable signal: locally both stacks produced ~6320 s and main had more >20 s clips (17 vs 14), even though CI showed dev consistently longer. The robust cross-environment signal is content mismatch, not duration.

## Remaining suspects (for the rebase delta)

The thinker decodes greedily (stage 0 `temperature: 0.0`), so the flip from "read the text" to "describe/answer the text" is a deterministic text-path perturbation. Consistent with this, Daily-Omni MCQ moves bit-stably 71.60% → 70.68% (11 single-token flips) at the same boundary.

1. **torch 2.13 compiled-region behavior** in the thinker — this branch already carries three torch-2.13 inductor workarounds (`acf245b2`, `e3b42850`, `de3502cf`).
2. **`e3efeaaf`** ("return a bare tensor from the qwen3-omni thinker when no capture is requested") — changing a `support_torch_compile`d forward's return arity can move the hidden-state capture relative to the compiled region; the tensors handed to the talker were never verified bit-identical.

Note the "good" dev nightlies 2942/2944/2946 actually ran a **main-lineage commit** (`900a7f081`), so no dev build before 2949 was ever benchmarked — the entire migration (vLLM 0.26→0.27, torch 2.11→2.13, ~30 rebase commits) lands in one unseparated window and nothing has been bisected yet.

## Suggested next steps

1. Bit-compare the thinker hidden states handed to the talker between the two stacks for one failing utterance (isolates suspect 2 from suspect 1).
2. Re-run the dev stack with `enforce_eager` on stage 0; if the extra failures disappear, it is inductor/torch 2.13.
3. Independently of the root cause: cap per-clip WER (or gate on the fraction of clips above a threshold) and seed the talker, so this gate stops being dominated by a handful of stochastic runaways.
4. Treat the task-text leakage (*"…speech in the same style as the reference audio"* spoken aloud, 64 clips even on main) as a product bug in its own right.

Related: #5480 (same job, earlier era where main itself sat at 0.40 with median 0.23 — distinct from this, where both branches have median 0.00 and differ only in tail mass).

**Comment by congw729** [comment]
Data point from the latest scheduled nightly on `main` (build [#13487](https://buildkite.com/vllm/vllm-omni/builds/13487), main @ `1ff27534`, 2026-08-12): the gate now fails on `main` itself, not only on the rebase branch:

```text
Mean WER:                                0.3840
- [Seed-TTS] seed_tts_content_error_mean (WER)=0.383989 > --max-seed-tts-mean-wer=0.35
FAILED tests/e2e/accuracy/qwen3_omni/test_qwen3_omni.py::test_qwen3_omni_seed_tts_wer_bench[omni_server0]
```

Step: [Omni · Accuracy Test](https://buildkite.com/vllm/vllm-omni/builds/13487#019ff1fe-74cf-4da1-b3c1-d7c12fa15152). Consistent with the latent failure mode documented here (main baseline 0.27-0.30 with uncapped WER outliers): a few instruction-following failures can tip the uncapped mean over the gate on any given night.

**Comment by congw729** [comment]
Update from the 2026-08-13 nightly (build [vllm-omni #13544](https://buildkite.com/vllm/vllm-omni/builds/13544), commit 5358f5db01e5):

`Omni · Accuracy Test` failed the seed-TTS WER gate again:

```text
Evaluated (WER, lower is better):        1088
Mean WER:                                0.3737
Median WER:                              0.0000
- [Seed-TTS] seed_tts_content_error_mean (WER)=0.373741 > --max-seed-tts-mean-wer=0.35
FAILED tests/e2e/accuracy/qwen3_omni/test_qwen3_omni.py::test_qwen3_omni_seed_tts_wer_bench[omni_server0] - assert 1 == 0
```

Mean WER 0.3737 remains above the 0.35 threshold, consistent with the regression tracked here.

**Comment by congw729** [comment]
Update from the 2026-08-14 nightly (build [vllm-omni #13582](https://buildkite.com/vllm/vllm-omni/builds/13582), commit 90a08c0edaa1):

`Omni · Accuracy Test` failed the seed-TTS WER gate again, and the number is getting worse:

```text
Mean WER:                                0.4333
- [Seed-TTS] seed_tts_content_error_mean (WER)=0.433280 > --max-seed-tts-mean-wer=0.35
FAILED tests/e2e/accuracy/qwen3_omni/test_qwen3_omni.py::test_qwen3_omni_seed_tts_wer_bench[omni_server0] - assert 1 == 0
```

Trend: 0.3737 (08-13) -> 0.4333 (08-14). This result is from a retried job (retries_count=1), so the regression reproduces across attempts.

- Step: https://buildkite.com/vllm/vllm-omni/builds/13582#019ffe08-cffc-4541-a460-8c8ec4d4fda6

**Comment by IneshReddy249** [comment]
Coming from #6090, which I'm assigned to — looks like the same regression through a different gate.

One thing that might be worth making explicit: #5976 merged to main on Aug 11, and build 13487 ran 1ff27534, which contains it. So main's jump to 0.3840 / 0.3737 / 0.4333 isn't main getting flaky — it's main now running the rebase stack. Pre-rebase main was 0.2665–0.2932 and post-rebase is 0.3737–0.4333, no overlap.

If nobody's on it, I'd like to take suggested step 3 — seeding the talker and capping per-clip WER. It's independent of the root cause and should stop this gate swinging on a few runaway clips. Leaving the rebase bisection to you.

**Comment by Gaohan123** [comment]
@ShengleiFu Could you please help take a look at the issue? Thanks

**Comment by ShengleiFu** [comment]
@Gaohan123 Sure, I can take a look, will share updates later. Thanks!

**Comment by Gaohan123** [comment]
> [@Gaohan123](https://github.com/Gaohan123) Sure, I can take a look, will share updates later. Thanks!

Thanks for your help!

**Comment by ShengleiFu** [comment]
<html><head></head><body><p><span>Root-cause update.</span></p><p><strong><span>TL;DR:</span></strong><span> the vLLM 0.27 / torch 2.13 regression localizes to the </span><strong><span>compiled stage-0 thinker path</span></strong><span>. A single-variable A/B — changing only stage-0 </span><code><span>enforce_eager</span></code><span> — moves the full Seed-TTS gate from fail back to the pre-rebase range. Disabling CUDA graphs alone does not recover accuracy, and the previously suspected bare-tensor return is not on the staged compiled path. The remaining evidence points to Inductor/numerical drift; </span><code><span>emulate_precision_casts</span></code><span> partially recovers it. A minimal upstream repro is still needed to determine whether the eventual fix belongs in vLLM or PyTorch.</span></p><p><span>Setup: 2×H800 (SM90, same arch as CI's H100), current main, </span><code><span>deploy ci/qwen3_omni_moe.yaml</span></code><span>, </span><code><span>seed-tts-eval en</span></code><span>. Only variable changed in the main A/B = stage-0 </span><code><span>enforce_eager</span></code><span>.</span></p><h3><span>1. Full-set A/B (N=1088)</span></h3>
stage-0 thinker | mean WER | median | WER > 1.0 | clips speaking "reference audio"
-- | -- | -- | -- | --
RED — compiled (default) | 0.4381 | 0.000 | 150 | 146
GREEN — enforce_eager | 0.2856 | 0.000 | 96 | 77

<p><span>RED reproduces the dev numbers here almost exactly (</span><code><span>151/146 -&gt; 150/146</span></code><span>). GREEN returns to the pre-rebase baseline (</span><code><span>~0.27–0.29</span></code><span>).</span></p><p><span>The only change between the two runs is stage-0 compilation mode.</span></p><h3><span>2. Bare-tensor suspect ruled out</span></h3><p><span>The inner forward wrapped by </span><code><span>@support_torch_compile</span></code><span> is byte-identical across the rebase.</span></p><p><span>The bare-tensor change is in the outer non-compiled forward, and that path is not exercised by staged execution: staged execution always captures intermediates and takes the tuple-return branch.</span></p><p><span>So this change does not explain the regression.</span></p><h3><span>3. Compile-knob bisect (N=128)</span></h3><ul><li><p><span>CUDA graph off: </span><code><span>0.4488</span></code></p></li><li><p><span>combo kernels off: </span><code><span>0.4331</span></code></p></li><li><p><code><span>emulate_precision_casts</span></code><span>: </span><code><span>0.3448</span></code></p></li><li><p><code><span>enforce_eager</span></code><span>: </span><code><span>0.2483</span></code></p></li></ul><p><span>Disabling CUDA graphs alone does not recover accuracy, so this is not a graph-capture issue.</span></p><p><code><span>emulate_precision_casts</span></code><span> recovers a substantial fraction of the gap, which points toward compile-time numerical behavior, but only full eager execution restores the baseline.</span></p><h3><span>4. Per-layer comparison</span></h3><p><span>For the same prompt, comparing compiled vs eager thinker activations:</span></p><ul><li><p><span>layer 0 is effectively identical;</span></p></li><li><p><span>divergence grows gradually through the transformer;</span></p></li><li><p><span>by layer 24 the relative divergence is ~22%;</span></p></li><li><p><span>final hidden state is ~12% relative divergence, with max absolute delta ~18.5.</span></p></li></ul><p><span>So I don't see a single abrupt layer boundary or isolated kernel that explains the regression. The numerical difference accumulates layer-by-layer.</span></p><h3><span>Conclusion</span></h3><p><span>The evidence localizes the rebase regression to numerical drift in the </span><strong><span>compiled stage-0 thinker path</span></strong><span>, rather than the vLLM-Omni staged-output plumbing.</span></p><p><span>For unblocking the gate, the reliable Omni-side workaround is to set stage 0 to </span><code><span>enforce_eager</span></code><span>. There is already precedent for scoped per-platform/per-stage eager overrides in the deploy configs.</span></p><p><span>I'll also measure the performance impact of stage-0 </span><code><span>enforce_eager</span></code><span> before proposing the workaround, and prepare a minimal upstream repro for vLLM/PyTorch. </span><code><span>emulate_precision_casts</span></code><span> looks like a useful lead for narrowing down the underlying Inductor/numerics issue.</span></p><p><span>The residual failures that remain under eager execution appear to be the pre-existing instruction-following failure mode tracked separately in #6090 / the original #6067 analysis.</span></p></body></html>

cc: @Gaohan123 @tzhouam PTAL and let me know if you have any suggestions, thanks!

**Comment by ShengleiFu** [comment]
Root-cause update + validation of the fix.

The v0.27 compile-induced regression this issue tracks is addressed by #6449 (thanks @tzhouam for the fix!), which pins the thinker's MRoPE behind the CUDA custom-op boundary so Inductor no longer vertically fuses the Q/K RMSNorm reductions into rotary. #6449's own validation covered the #6090 5-request audio gate; below is the missing piece — validation against the **full 1088-clip Seed-TTS WER gate this issue is gated on**, on 2×H800 (SM90, same architecture as CI's H100).

**Setup:** vLLM 0.27.0 / torch 2.13+cu130, `Qwen/Qwen3-Omni-30B-A3B-Instruct`, `seed-tts-eval` en (N=1088). Single variable = stage-0 `compilation_config.custom_ops: ["+rotary_embedding"]` (#6449) layered on the CI config.

| stage-0 thinker | mean WER ↓ | WER>1.0 | "reference audio" leak |
| --- | ---: | ---: | ---: |
| compiled (default) — stored | 0.4381 | 150 | 146 |
| compiled (default) — fresh anchor | 0.3970 | 153 | 145 |
| `enforce_eager` (oracle) | 0.2856 | 96 | 77 |
| **`+rotary_embedding` (#6449)** | **0.1815** | **48** | **29** |

- #6449 passes the 0.35 gate, at/below the `enforce_eager` oracle on this aggregate metric, while keeping compilation enabled.
- Of the 84 clips where the default compiled path regresses relative to eager (`WER_compiled >= 0.5 && WER_eager < 0.5`), #6449 recovers **81/84 (96%)**. On that subset: default compiled WER **2.25** → `+rotary_embedding` **0.08**, close to eager at **0.045**.
- The fresh compiled anchor still reproduces the regression in the same environment and recovers only **3/84**, which makes environment drift unlikely and supports the recovery being caused by the change.

**Example recovered clips (ASR of generated audio):**

- ref: *"Cribb was born near Bristol…"* — default compiled babbles *"Minni, the one who is the one who is…"* (WER 34) → #6449 reads the reference correctly.
- ref: *"Appearances are deceptive."* — default compiled speaks the task description *"emphasizing speech in the same language and style as the reference audio."* → #6449 reads the reference correctly.

**Residual:** ~48 clips with WER>1.0 / 29 reference-audio leaks remain, and the same failure mode is present with `enforce_eager` too — i.e. the pre-existing instruction-following/product issue rather than the v0.27 compile-induced regression.

Suggested tracking: treat the **compile-induced regression as resolved by #6449**, and track the residual eager-reproducible instruction-following failures separately (aligned with the #6090 metric-hardening work).

@tzhouam plz let me know if you have suggestions here. Thanks!

**Comment by Gaohan123** [comment]
Thanks for the analysis from @ShengleiFu . The issue is solved by PR #6449
